wgpu/_compute.py

```diff
@@ -0,0 +1,152 @@
+"""
+Simple high-level utilities for doing compute on the GPU.
+
+The purpose of wgpu-py is to provide a Pythonic wrapper around wgpu-native.
+In principal, a higher-level API is not within the scope of the project.
+However, by providing a few un-opinionated utility functions, other projects
+(like python-shader) can use wgpu (e.g. in their tests) without having
+to keep track of changes in wgpu itself.
+"""
+
+import sys
+
+import wgpu
+
+# Note: these funcs need numpy, but we don't import it because its not a strict dep of wgpu-py
+
+
+def compute_with_buffers(input_arrays, output_arrays, shader, n=None):
+    """ Apply the given compute shader to the given input_arrays and return
+    output arrays. Both input and output arrays are represented on the GPU
+    using storage buffer objects.
+
+    Params:
+        input_arrays (dict): A dict mapping an integer binding to a numpy array.
+        output_arrays (dict): A dict mapping an integer binding to an outputs spec.
+            The spec must be a tuple ``(*shape, dtype)``. A buffer is created
+            for each binding and its data is read and returned. The binding can be
+            one that is also used as an input array (in which case the GPU buffer
+            is the same).
+        shader (bytes, shader-object): Any compatible shader object representing
+            a SpirV compute shader.
+        n (int, tuple, optional): The number provided to dispatch. Can be an int
+            or a 3-tuple of ints to specify (x, y, z). If not given or None,
+            the first value of the first spec in output_arrays is used.
+    """
```

Comment:
@Korijn do you have any thoughts about how to represent these "output specs"? The idea is that we need to know the shape and dtype so that we a) know the size of the buffer, and b) can "cast" the data before returning it.

Reply:
I think what you've got going on now is pretty good! I just wonder if it needs to be a dict if the keys are incrementing integers? It could just be a list or tuple right?

Reply:
Because (I think) the numbers don't *have* to be incremental. And also to allow specifying the same index for an input and output.

Dynd has this nice feature that the dtype includes the shape, e.g. `10x10xfloat32`. That would have been nice here, but Numpy does not have such a thing, does it?

Reply:
No, there is a project going on to add that to the numpy typeshed but I don't believe they ever got that to work... I don't think it's a huge deal though.

Reply:
Here it is:  https://github.com/numpy/numpy-stubs